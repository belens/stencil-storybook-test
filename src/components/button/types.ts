import type { ButtonValues } from '@components/atoms/button/values';
import type { IconValues } from '@components/atoms/icon/values';

export namespace ButtonTypes {
  export type Size = `${ButtonValues.Size}`;
  export type Type = `${ButtonValues.Type}`;
  export type Variant = `${ButtonValues.Variant}`;

  export interface Props {
    text: string;
    name: string;
    type: ButtonValues.Type;
    disabled: boolean;
    variant?: ButtonValues.Variant;
    icon?: IconValues.Name;
    size?: ButtonValues.Size;
  }
}
