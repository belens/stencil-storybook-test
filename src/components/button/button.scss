mac-button {
	display: inline-block;
}

.mac-button {
	display: inline-flex;
	flex-direction: row;
	align-items: center;
	justify-content: center;
	gap: 10px;
	width: var(--button-width, auto);
	padding: 7px 11px;
	border: 1px solid transparent;
	border-radius: 6px;
	background-color: var(--color-neutral-100);
	color: var(--text-color);
	font-family: inherit;
	font-size: 1rem;
	font-weight: 600;
	line-height: 1.5;
	text-align: center;
	user-select: none;
	cursor: pointer;
	transition:
		border-color .15s ease-in-out,
		background-color .15s ease-in-out,
		color .15s ease-in-out,
		box-shadow .15s ease-in-out
	;
	vertical-align: top;
	white-space: nowrap;

	&:focus {
		color: var(--color-neutral-800);
		outline: 0;
		box-shadow: 0 2px 4px -2px rgba($box-shadow-color, .1), 0 4px 6px -1px rgba($box-shadow-color, .1);
	}

	&:hover,
	&:active {
		background-color: var(--color-neutral-200);
		color: var(--color-neutral-800);
	}

	&:disabled {
		background-color: var(--color-neutral-100);
		color: var(--color-neutral-300);
		cursor: not-allowed;
	}

	&--primary {
		background-color: var(--color-primary-700);
		color: #fff;

		&:focus {
			background-color: var(--color-primary-500);
			color: #fff;
		}

		&:hover,
		&:active {
			background-color: var(--color-primary-900);
			color: #fff;
		}

		&:disabled {
			background-color: var(--color-neutral-100);
			color: var(--color-neutral-300);
		}
	}

	&--flat {
		background-color: #fff;

		&:focus {
			background-color: #fff;
		}

		&:disabled {
			background-color: var(--color-neutral-100);
		}
	}

	&--line {
		border-color: var(--color-neutral-200);
		background-color: #fff;

		&:focus {
			border-color: var(--color-neutral-300);
			background-color: #fff;
			box-shadow: unset;
		}

		&:hover,
		&:active {
			border-color: var(--color-neutral-300);
		}

		&:disabled {
			border-color: var(--color-neutral-200);
			background-color: #fff;
		}
	}

	&--destructive {
		background-color: var(--color-error-700);
		color: #fff;

		&:focus,
		&:hover,
		&:active {
			background-color: var(--color-error-900);
			color: #fff;
		}

		&:disabled {
			background-color: var(--color-error-50);
			color: var(--color-error-300);
		}
	}

	&--icon-only {
		min-width: unset;
		padding: 10px;
	}

	&--small {
		gap: 6px;
		padding: 3px 7px;
		font-size: .875rem;
		line-height: 1.4285714286;
	}

	&--icon-only#{&}--small {
		padding: 5px;
	}

	&--large {
		padding: 9px 15px;
		font-size: 1.125rem;
		line-height: 1.555556;
	}

	&--icon-only#{&}--large {
		padding: 12px;
	}
}
