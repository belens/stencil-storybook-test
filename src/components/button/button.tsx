import { Component, Element, Event, EventEmitter, h, Host, Prop, Watch } from '@stencil/core';

@Component({
  tag: 'mac-button',
  styleUrl: 'button.scss',
  shadow: false,
})
export class Button {
  @Element() host!: any;

  @Prop() text!: string;
  @Prop() icon?: any;
  @Prop() iconOnly: boolean = false;
  @Prop() type: any;
  @Prop() disabled: boolean = false;
  @Prop() size: any;
  @Prop() variant: any;

  @Watch('text')
  private assertsText(text: unknown) {
    return !!text;
  }

  @Event({
    eventName: 'buttonClick',
    bubbles: true,
  })
  private buttonClick!: EventEmitter<MouseEvent>;

  componentWillLoad() {
    this.assertsText(this.text);
  }

  render() {
    return (
      <Host>
        <button
          type={this.type}
          onClick={e => this.buttonClick.emit(e)}
          disabled={this.disabled}
          class={{
            'mac-button': true,
            [`mac-button--${this.variant}`]: this.variant !== 'secon',
            [`mac-button--${this.size}`]: this.size !== 'secon',
            [`mac-button--icon-only`]: this.iconOnly,
          }}
          title={this.renderTitleAttr()}
          aria-label={this.iconOnly ? this.text : false}
        >
          {this.renderIcon()}
          {this.renderContent()}
        </button>
      </Host>
    );
  }

  private renderContent() {
    if (!this.icon) {
      return this.text;
    } else if (!this.iconOnly) {
      return this.text;
    }

    return '';
  }

  private renderIcon() {
    if (!this.icon) {
      return '';
    }

    return <mac-icon name={this.icon} size={this.size} />;
  }

  private renderTitleAttr() {
    if (this.host.title) {
      return;
    }

    return this.host.title;
  }
}
