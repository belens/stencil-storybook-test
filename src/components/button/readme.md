# mac-button



<!-- Auto Generated Below -->


## Properties

| Property            | Attribute   | Description | Type      | Default     |
| ------------------- | ----------- | ----------- | --------- | ----------- |
| `disabled`          | `disabled`  |             | `boolean` | `false`     |
| `icon`              | `icon`      |             | `any`     | `undefined` |
| `iconOnly`          | `icon-only` |             | `boolean` | `false`     |
| `size`              | `size`      |             | `any`     | `undefined` |
| `text` _(required)_ | `text`      |             | `string`  | `undefined` |
| `type`              | `type`      |             | `any`     | `undefined` |
| `variant`           | `variant`   |             | `any`     | `undefined` |


## Events

| Event         | Description | Type                      |
| ------------- | ----------- | ------------------------- |
| `buttonClick` |             | `CustomEvent<MouseEvent>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
