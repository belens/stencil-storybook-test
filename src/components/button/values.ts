export namespace ButtonValues {
  export enum Size {
    SMALL = 'small',
    MEDIUM = 'medium',
    LARGE = 'large',
  }

  export enum Variant {
    PRIMARY = 'primary',
    SECONDARY = 'secondary',
    FLAT = 'flat',
    LINE = 'line',
    DESTRUCTIVE = 'destructive',
  }

  export enum Type {
    BUTTON = 'button',
    SUBMIT = 'submit',
    RESET = 'reset',
  }
}
