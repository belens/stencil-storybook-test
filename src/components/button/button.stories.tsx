import { h } from '@stencil/core';
import { ButtonValues } from './values';
// import { IconValues } from '../icon/values';

const names = ['s'];
const labels = [''];

export default {
  title: 'Atoms/button',
  argTypes: {
    text: {
      control: { type: 'text' },
    },
    icon: {
      options: names,
      control: { type: 'select', labels },
    },
    iconOnly: {
      control: { type: 'boolean' },
    },
    variant: {
      control: { type: 'inline-radio' },
      options: Object.values(ButtonValues.Variant),
    },
    size: {
      control: { type: 'inline-radio' },
      options: Object.values(ButtonValues.Size),
    },
    title: {
      control: { type: 'text' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    onButtonClick: { action: 'onButtonClick' },
  },
  args: {
    text: 'Button text',
    iconOnly: false,
    variant: ButtonValues.Variant.SECONDARY,
    size: ButtonValues.Size.MEDIUM,
    disabled: false,
  },
};

export const Default = args => {
  return <div>asdf</div>;
};
