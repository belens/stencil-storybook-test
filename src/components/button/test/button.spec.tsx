import { ButtonValues } from '@components/atoms/button/values';
import { IconValues } from '@components/atoms/icon/values';
import { newSpecPage } from '@stencil/core/testing';

import { Button } from '../button';

// only run this file --> npx stencil test --spec -- button.spec.tsx
describe('<mac-button />', () => {
  it('should render with default properties', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<mac-button text="Click me" />',
    });

    expect(page.root).toMatchSnapshot();
  });

  it('should have a title', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<mac-button text="Click me" title="You should click me" />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it.each(Object.values(ButtonValues.Variant))('should add css class for variant %s', async variant => {
    const page = await newSpecPage({
      components: [Button],
      html: `<mac-button text="Click me" variant="${variant}" />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it.each(Object.values(ButtonValues.Size))('should add css class for size %s', async size => {
    const page = await newSpecPage({
      components: [Button],
      html: `<mac-button text="Click me" size="${size}" />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('should be disabled', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<mac-button text="Click me" disabled="" />',
    });

    expect(page.root).toMatchSnapshot();
  });

  it('should be a submit button', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<mac-button type="submit" text="Click me" />',
    });

    expect(page.root).toMatchSnapshot();
  });

  it('should have an icon', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<mac-button text="Click me" icon="${IconValues.Name.ADD}" />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('should only have an icon (and an aria-label)', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<mac-button text="Click me" icon="${IconValues.Name.ADD}" icon-only="" />`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
