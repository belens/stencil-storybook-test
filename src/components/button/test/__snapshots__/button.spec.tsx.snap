// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<mac-button /> should add css class for size large 1`] = `
<mac-button size="large" text="Click me">
  <button class="mac-button mac-button--large" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should add css class for size medium 1`] = `
<mac-button size="medium" text="Click me">
  <button class="mac-button" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should add css class for size small 1`] = `
<mac-button size="small" text="Click me">
  <button class="mac-button mac-button--small" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should add css class for variant destructive 1`] = `
<mac-button text="Click me" variant="destructive">
  <button class="mac-button mac-button--destructive" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should add css class for variant flat 1`] = `
<mac-button text="Click me" variant="flat">
  <button class="mac-button mac-button--flat" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should add css class for variant line 1`] = `
<mac-button text="Click me" variant="line">
  <button class="mac-button mac-button--line" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should add css class for variant primary 1`] = `
<mac-button text="Click me" variant="primary">
  <button class="mac-button mac-button--primary" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should add css class for variant secondary 1`] = `
<mac-button text="Click me" variant="secondary">
  <button class="mac-button" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should be a submit button 1`] = `
<mac-button text="Click me" type="submit">
  <button class="mac-button" type="submit">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should be disabled 1`] = `
<mac-button disabled="" text="Click me">
  <button class="mac-button" disabled="" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should have a title 1`] = `
<mac-button text="Click me" title="You should click me">
  <button class="mac-button" title="You should click me" type="button">
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should have an icon 1`] = `
<mac-button icon="add" text="Click me">
  <button class="mac-button" type="button">
    <mac-icon name="add" size="medium"></mac-icon>
    Click me
  </button>
</mac-button>
`;

exports[`<mac-button /> should only have an icon (and an aria-label) 1`] = `
<mac-button icon="add" icon-only="" text="Click me">
  <button aria-label="Click me" class="mac-button mac-button--icon-only" type="button">
    <mac-icon name="add" size="medium"></mac-icon>
  </button>
</mac-button>
`;

exports[`<mac-button /> should render with default properties 1`] = `
<mac-button text="Click me">
  <button class="mac-button" type="button">
    Click me
  </button>
</mac-button>
`;
